src/shims/sync.rs

```diff
@@ -0,0 +1,581 @@
+    }
+
+    fn pthread_rwlock_trywrlock(&mut self, rwlock_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let rwlock = this.read_scalar(rwlock_op)?.not_undef()?;
+        if this.is_null(rwlock)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+
+        let readers = rwlock_get_readers(this, rwlock_op)?.to_u32()?;
+        let writers = rwlock_get_writers(this, rwlock_op)?.to_u32()?;
+        if readers != 0 || writers != 0 {
+            this.eval_libc_i32("EBUSY")
+        } else {
+            rwlock_set_writers(this, rwlock_op, Scalar::from_u32(1))?;
+            Ok(0)
+        }
+    }
+
+    fn pthread_rwlock_unlock(&mut self, rwlock_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let rwlock = this.read_scalar(rwlock_op)?.not_undef()?;
+        if this.is_null(rwlock)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+
+        let readers = rwlock_get_readers(this, rwlock_op)?.to_u32()?;
+        let writers = rwlock_get_writers(this, rwlock_op)?.to_u32()?;
+        if let Some(new_readers) = readers.checked_sub(1) {
+            rwlock_set_readers(this, rwlock_op, Scalar::from_u32(new_readers))?;
+            Ok(0)
+        } else if writers != 0 {
+            rwlock_set_writers(this, rwlock_op, Scalar::from_u32(0))?;
+            Ok(0)
+        } else {
+            this.eval_libc_i32("EPERM")
+        }
+    }
+
+    fn pthread_rwlock_destroy(&mut self, rwlock_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        let rwlock = this.read_scalar(rwlock_op)?.not_undef()?;
+        if this.is_null(rwlock)? {
+            return this.eval_libc_i32("EINVAL");
+        }
+
+        if rwlock_get_readers(this, rwlock_op)?.to_u32()? != 0 {
+            return this.eval_libc_i32("EBUSY");
+        }
+        if rwlock_get_writers(this, rwlock_op)?.to_u32()? != 0 {
+            return this.eval_libc_i32("EBUSY");
+        }
+
+        rwlock_set_readers(this, rwlock_op, ScalarMaybeUndef::Undef)?;
+        rwlock_set_writers(this, rwlock_op, ScalarMaybeUndef::Undef)?;
+
+        Ok(0)
+    }
+}
+
+fn assert_ptr_target_min_size<'mir, 'tcx: 'mir>(
+    ecx: &MiriEvalContext<'mir, 'tcx>,
+    operand: OpTy<'tcx, Tag>,
+    min_size: u64,
+) -> InterpResult<'tcx, ()> {
+    let target_ty = match operand.layout.ty.kind {
+        TyKind::RawPtr(TypeAndMut { ty, mutbl: _ }) => ty,
+        _ => panic!("Argument to pthread function was not a raw pointer"),
+    };
+    let target_layout = ecx.layout_of(target_ty)?;
+    assert!(target_layout.details.size.bytes() >= min_size);
+    Ok(())
+}
+
+// pthread_mutexattr_t is either 4 or 8 bytes, depending on the platform.
+
+// Our chosen memory layout: store an i32 in the first four bytes equal to the
+// corresponding libc mutex kind constant (i.e. PTHREAD_MUTEX_NORMAL)
+
+fn mutexattr_get_kind<'mir, 'tcx: 'mir>(
```

Comment:
Please move these helper functions up above. Generally when reading code top-to-bottom (in particular for review in GH) I find it much easier to follow if things are defined before they are called.